src/cache.ts

```diff
@@ -0,0 +1,221 @@
+			return deserialize(val);
+		}
+		const val = await setValFn(key);
+		await cache.put(
+			request,
+			new Response(serialize(val.value), {
+				headers: { expires: val.expiration.toUTCString() },
+			})
+		);
+		return val.value;
+	}
+}
+
+export class MixedCache implements ReadableCache {
+	private caches: ReadableCache[];
+	constructor(...caches: ReadableCache[]) {
+		this.caches = caches;
+	}
+
+	async read<T>(key: string, setValFn: (key: string) => Promise<CacheElement<T>>): Promise<T> {
+		const caches = this.caches;
+		const setValFnBuilder = (i: number): ((key: string) => Promise<CacheElement<T>>) => {
+			if (i >= caches.length) {
+				return setValFn;
+			}
+			return async (key: string) => ({
+				value: await caches[i].read(key, setValFnBuilder(i + 1)),
+				expiration: new Date(), // TODO: find a way to use setVal expiration instead
+			});
+		};
+		const cachedValue = await setValFnBuilder(0)(key);
+		return cachedValue.value;
+	}
+}
+
+export const DEFAULT_R2_BUCKET_CACHE_TTL_IN_MS = 5 * 60 * 1000;
+
+export class CachedR2Object {
+	public checksums: R2Checksums;
+	public customMetadata?: Record<string, string>;
+	public etag: string;
+	public httpEtag: string;
+	public httpMetadata?: R2HTTPMetadata;
+	public key: string;
+	public size: number;
+	public uploaded: Date;
+	public version: string;
+
+	constructor(
+		object: R2Object,
+		public body?: Uint8Array
+	) {
+		this.checksums = object.checksums;
+		this.customMetadata = object.customMetadata;
+		this.etag = object.etag;
+		this.httpEtag = object.httpEtag;
+		this.httpMetadata = object.httpMetadata;
+		this.key = object.key;
+		this.size = object.size;
+		this.uploaded = object.uploaded;
+		this.version = object.version;
+	}
+}
+
+export class CachedR2Objects {
+	public delimitedPrefixes: string[];
+	public objects: CachedR2Object[];
+	public truncated: boolean;
+
+	constructor(objects: R2Objects) {
+		this.delimitedPrefixes = objects.delimitedPrefixes;
+		this.objects = objects.objects.map(o => new CachedR2Object(o));
+		this.truncated = objects.truncated;
+	}
+}
+
+const R2Method = {
+	GET: 'get',
+	HEAD: 'head',
+	LIST: 'list',
+};
+
+export class CachedR2Bucket {
+	constructor(
+		private ctx: Context,
+		private bucket: R2Bucket,
+		private cache: ReadableCache,
+		private ttl_in_ms = DEFAULT_R2_BUCKET_CACHE_TTL_IN_MS
+	) {}
+	head(key: string): Promise<CachedR2Object | null> {
+		const cacheKey = `head/${key}`;
+		return this.cache.read(cacheKey, async () => {
+			this.ctx.metrics.r2RequestsTotal.inc({ method: R2Method.HEAD });
+			const object = await this.bucket.head(key);
+			if (object === null) {
+				return { value: null, expiration: new Date() };
+			}
+			const value = new CachedR2Object(object);
+			return {
+				value,
+				expiration: new Date(Date.now() + this.ttl_in_ms),
+			};
+		});
+	}
+
+	list(options?: R2ListOptions | undefined): Promise<CachedR2Objects> {
+		const cacheKey = `list/${JSON.stringify(options)}`;
```

Comment:
Objects keys are limited to 1024 according to https://developers.cloudflare.com/r2/reference/limits/

Are there sufficient checks in place such that the cache keys for the head, list, and get methods are not hitting that limit?

The users below seems to be fine, but maybe if this gets copied and pasted elsewhere, it might not satisfy the requirement. Perhaps a documentation comment would be good?

Review of cache use below:

OK, `truncatedTokenKeyId` is a number:
```js
	const key = await ctx.cache.ISSUANCE_KEYS.get(tokenRequest.truncatedTokenKeyId.toString());
```

OK, the list options are hard-coded:
```js
	const keys = await ctx.cache.ISSUANCE_KEYS.list({ include: ['customMetadata'] });
```


OK, `tokenKeyID` is the SHA256 hash from `keyToTokenKeyID`:
```js
	} while ((await ctx.cache.ISSUANCE_KEYS.head(tokenKeyID.toString())) !== null);
```

Reply:
The object key is not modified. The only key that's instantiated is the cache key.
I've added a comment in case this code is copy pasted and modified.